### PR TITLE
Feature/10257 fss UI unit accessibility functionaltests

### PR DESCRIPTION
**Pull request by Rashi14748**
Fss UI Home Screen : Added Accessibility tests & Functional tests.
                                 : Updated folder structure for Unit Tests.
                                 :  transformation changes have been added in azure-pipelines.yml

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -20,7 +20,15 @@ Run `ng build` to build the project. The build artifacts will be stored in the `
 
 ## Running unit tests
 
-Run `npm test` to execute the unit tests via [Jest](https://jestjs.io/).
+Run `npm run test:unit` to execute the unit tests via [Jest](https://jestjs.io/).
+
+## Running functional tests
+
+Run `npm run test:auto` to execute the unit tests via [Jest-Playwright].
```
Comment (gregzealley) [inline README.md:27]:
```suggestion
Run `npm run test:auto` to execute the unit tests via [Jest-Playwright](https://www.npmjs.com/package/jest-playwright-preset).
```

File: README.md [hunk 2]
```diff
@@ -20,7 +20,15 @@ Run `ng build` to build the project. The build artifacts will be stored in the `
...
+
+## Running functional tests
+
+Run `npm run test:a11y` to execute the unit tests via [Jest-Axe].
```
Comment (gregzealley) [inline README.md:31]:
```suggestion
Run `npm run test:a11y` to execute the unit tests via [Jest-Axe](https://www.npmjs.com/package/jest-axe).
```

File: Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    beforeEach(async () => {
+      await TestBed.configureTestingModule({
+        imports: [RouterTestingModule],
+        declarations: [FssFooterComponent , FooterComponent],
+        schemas: [NO_ERRORS_SCHEMA]
+      }).compileComponents();     
+      
+    });
+
+    test('should retuns no violations for ukho-footer component', async() =>{
```
Comment (gregzealley) [inline Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts:21]:
```suggestion
    test('should return no violations for ukho-footer component', async() =>{
```

File: Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        const fixture = TestBed.createComponent(FssFooterComponent);
+        fixture.detectChanges();     
+        const footer = fixture.debugElement.query(By.css('ukho-footer'));
+        const results= await axe(footer.nativeElement);
+        expect(results.violations).toEqual([]);
+      });
+
+      test('should retuns no violation for image in ukho-footer component', async() =>{
```
Comment (gregzealley) [inline Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts:29]:
```suggestion
      test('should return no violation for image in ukho-footer component', async() =>{
```

File: Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        const fixture = TestBed.createComponent(FssFooterComponent);
+        fixture.detectChanges();     
+        const footer = fixture.debugElement.query(By.css('ukho-footer')).childNodes;
+        const imglink=footer[0].nativeNode.querySelector('img');        
+        const results= await axe(imglink);
+        expect(results.violations).toEqual([]);      
+      });
+
+      test('should retuns no violations for Privacy policy anchor tag in ukho-footer', async() =>{
```
Comment (gregzealley) [inline Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts:38]:
```suggestion
      test('should return no violations for Privacy policy anchor tag in ukho-footer', async() =>{
```

File: Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        const fixture = TestBed.createComponent(FssFooterComponent);
+        fixture.detectChanges();     
+        const footer = fixture.debugElement.query(By.css('ukho-footer')).childNodes;        
+        const link= footer[1].nativeNode.querySelector('a');
+        expect(link.textContent).toEqual("Privacy policy");
+        const results= await axe(link);
+        expect(results.violations).toEqual([]);
+      });
+
+      test('should retuns no violations for Accessibility anchor tag in ukho-footer', async() =>{
```
Comment (gregzealley) [inline Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts:48]:
```suggestion
      test('should return no violations for Accessibility anchor tag in ukho-footer', async() =>{
```

File: Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        const fixture = TestBed.createComponent(FssFooterComponent);
+        fixture.detectChanges();     
+        const footer = fixture.debugElement.query(By.css('ukho-footer')).childNodes;
+        const link= footer[1].nativeNode.querySelector('a:nth-child(2)');
+        expect(link.textContent).toEqual("Accessibility");
+        const results = await axe(link);
+        expect(results.violations).toEqual([]);   
+      });
+
+      test('should retuns pass as true for all anchor tag href attribute present in ukho-footer', async() =>{
```
Comment (gregzealley) [inline Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts:58]:
```suggestion
      test('should return pass as true for all anchor tag href attribute present in ukho-footer', async() =>{
```

File: Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        const fixture = TestBed.createComponent(FssFooterComponent);
+        fixture.detectChanges();     
+        const footeranchortags = fixture.debugElement.queryAll(By.css('a'));
+        for (var i = 0; i < footeranchortags.length; i++) {
+            var anchortags = footeranchortags[i];
+            expect(anchortags.attributes['href']).not.toBe(null);            
+        }           
+      });      
+      
+      test('should returns violations for mock html', async() =>{
```
Comment (gregzealley) [inline Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts:68]:
```suggestion
      test('should return violations for mock html', async() =>{
```

File: Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+                                    <html>
+                                        <body>
+                                           <a href="#"></a>
+                                           <img src="#"/>
+                                        </body>
+                                    </html>`
+        const render = () => failingExtendedHtmlExample
+        const html = render()                            
+        const results = await axe(html)
+        expect(results.violations.length).toBeGreaterThan(0)
```
Comment (gregzealley) [inline Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts:79]: Is there a reason this assert is different to the others (expect(results.violations).toEqual([]);) ?

File: Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+      test('should return no violation for "Skip to content" in an anchor tag', async() =>{
+        const fixture = TestBed.createComponent(FssHeaderComponent);
+        fixture.detectChanges();   
+        const anchortag = fixture.debugElement.query(By.css('a')).nativeElement;  
+        expect(anchortag.textContent).toContain('Skip to content');
+        const results= await axe(anchortag)
+        expect(results.violations).toEqual([])
+        });
+
+        test('should return pass as true for href attribute present in an anchor tag', async() =>{
```
Comment (gregzealley) [inline Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts:29]:
```suggestion
        test('should return true for href attribute present in an anchor tag', async() =>{
```

File: Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+        test('should return no violation for ukho-header component ', async() =>{
+            const fixture = TestBed.createComponent(FssHeaderComponent);
+            fixture.detectChanges();   
+            const header =fixture.debugElement.query(By.css('ukho-header'));            
+            const results= await axe(header.nativeElement);
+            expect(results.violations).toEqual([]);
+            });
+
+        test('should returns no violation for branding logo under ukho-header component ', async() =>{
```
Comment (gregzealley) [inline Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts:44]:
```suggestion
        test('should return no violation for branding logo under ukho-header component ', async() =>{
```

File: Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+                const fixture = TestBed.createComponent(FssHeaderComponent);
+                fixture.detectChanges();   
+                const headernodes =fixture.debugElement.query(By.css('ukho-header')).childNodes;            
+                const imglink=headernodes[0].nativeNode.querySelector('a');        
+                const results= await axe(imglink);
+                expect(results.violations).toEqual([]);      
+        });
+
+        test('should returns no violation for image logo under ukho-header component ', async() =>{
```
Comment (gregzealley) [inline Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts:53]:
```suggestion
        test('should return no violation for image logo under ukho-header component ', async() =>{
```

File: Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+          const fixture = TestBed.createComponent(FssHeaderComponent);
+          fixture.detectChanges();   
+          const headernodes =fixture.debugElement.query(By.css('ukho-header')).childNodes;            
+          const imglink=headernodes[0].nativeNode.querySelector('img');        
+          const results= await axe(imglink);
+          expect(results.violations).toEqual([]);      
+        });
+
+
+        test('should return pass as false for no href attribute search link under ukho-header component ', async() =>{
```
Comment (gregzealley) [inline Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts:63]:
```suggestion
        test('should return false for no href attribute search link under ukho-header component ', async() =>{
```

File: Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+            const fixture = TestBed.createComponent(FssHeaderComponent);
+            fixture.detectChanges();   
+            const allanchorrnodes =fixture.debugElement.queryAll(By.css('a')); 
+            const linksearch=allanchorrnodes[2];      
+            expect(linksearch.nativeElement.textContent).toEqual('Search');         
+            expect(linksearch.attributes['href']).toBeFalsy();   
+        });
+
+        test('should return pass as false for no href attribute Sign In link under ukho-header component ', async() =>{
```
Comment (gregzealley) [inline Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts:72]:
```suggestion
        test('should return false for no href attribute Sign In link under ukho-header component ', async() =>{
```

File: Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+          const fixture = TestBed.createComponent(FssHeaderComponent);
+          fixture.detectChanges();   
+          const allanchorrnodes =fixture.debugElement.queryAll(By.css('a')); 
+          const linksearch=allanchorrnodes[3];      
+          expect(linksearch.nativeElement.textContent).toEqual('Sign in');         
+          expect(linksearch.attributes['href']).toBeFalsy();   
+      });
+
+      test('should returns no violation for header under ukho-header component ', async() =>{
```
Comment (gregzealley) [inline Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts:81]:
```suggestion
      test('should returnsno violation for header under ukho-header component ', async() =>{
```

File: Tests/FunctionalTests/fss-home.spec.ts [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+          }) 
+    })
+
+    test('Does it navigate to marine data portal page once click on marine data portal link', async () => {
+        await page.click(pageObjectsConfig.marinedataportallinkSelector);
+        expect(await page.innerText(pageObjectsConfig.marinedataportalpageSelector)).toContain(pageObjectsConfig.marinedataportalpageText);
+        expect(page.url()).toContain("marine-data-portal");
+    }) 
+
+    test('Does it navigate to admiralty home page once click on UK Hydrographic office link', async () => {
```
Comment (gregzealley) [inline Tests/FunctionalTests/fss-home.spec.ts:62]:
```suggestion
    test('Does it navigate to Admiralty home page once click on UK Hydrographic Office link', async () => {
```

File: Tests/FunctionalTests/fss-home.spec.ts [hunk 17]
```diff
@@ -0,0 +1,69 @@
+import { Browser, BrowserContext, chromium, Page } from 'playwright';
```
Comment (gregzealley) [inline Tests/FunctionalTests/fss-home.spec.ts:1]: Do we need  a test for the Terms of Use link ?

File: appConfig.ts [hunk 18]
```diff
@@ -2,5 +2,6 @@
 export const fssConfiguration = {
     copyright: "@ Crown copyright 2021 UK Hydrographic office",
```
Comment (gregzealley) [inline appConfig.ts:3]:
```suggestion
    copyright: "@ Crown copyright 2021 UK Hydrographic Office",
```

**Review by gregzealley — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (Rashi14748) [reply README.md:27]: Thanks, updated Read me with the provided link

File: README.md (hunk 2 above)
Comment (Rashi14748) [reply README.md:31]: Thanks, updated Read me with the provided link

File: Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts (hunk 3 above)
Comment (Rashi14748) [reply Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts:21]: updated

File: Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts (hunk 4 above)
Comment (Rashi14748) [reply Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts:29]: updated

File: Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts (hunk 5 above)
Comment (Rashi14748) [reply Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts:38]: updated

File: Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts (hunk 6 above)
Comment (Rashi14748) [reply Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts:48]: updated

File: Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts (hunk 7 above)
Comment (Rashi14748) [reply Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts:58]: updated

File: Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts (hunk 8 above)
Comment (Rashi14748) [reply Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts:68]: updated

File: Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts (hunk 10 above)
Comment (Rashi14748) [reply Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts:29]: updated

File: Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts (hunk 11 above)
Comment (Rashi14748) [reply Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts:44]: updated

File: Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts (hunk 12 above)
Comment (Rashi14748) [reply Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts:53]: updated

File: Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts (hunk 13 above)
Comment (Rashi14748) [reply Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts:63]: updated

File: Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts (hunk 14 above)
Comment (Rashi14748) [reply Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts:72]: updated

File: Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts (hunk 15 above)
Comment (Rashi14748) [reply Tests/AccessibilityTests/accessibility-fss-home-page-header.spec.ts:81]: updated

File: Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts (hunk 9 above)
Comment (kajalkundu) [reply Tests/AccessibilityTests/accessibility-fss-home-page-footer.spec.ts:79]: This test I added with fake html element and verified that it should returned violations. That's why verified that count should be more than zero.

File: Tests/FunctionalTests/fss-home.spec.ts (hunk 16 above)
Comment (Rashi14748) [reply Tests/FunctionalTests/fss-home.spec.ts:62]: updated

File: appConfig.ts (hunk 18 above)
Comment (Rashi14748) [reply appConfig.ts:3]: updated

File: Tests/FunctionalTests/fss-home.spec.ts (hunk 17 above)
Comment (kajalkundu) [reply Tests/FunctionalTests/fss-home.spec.ts:1]: Hi Greg, There is no "Terms of Use link" on FSS UI therefore no test added for the same.

**Review by bhalcha13813 — APPROVED** [review]
LGTM

**Review by gregzealley — APPROVED** [review]

**Review by martyn-fewtrell — APPROVED** [review]
Looks fine

**Pull request merged**